packages/typespec-client-generator-core/multipart_api_design.md

```diff
@@ -0,0 +1,84 @@
+This doc is to design TCGC API of multipart for language emitters.
+
+# multipart payload
+
+Here is classic payload of multipart:
+```
+POST /upload HTTP/1.1
+Content-Length: 428
+Content-Type: multipart/form-data; boundary=abcde12345
+--abcde12345
+Content-Disposition: form-data; name="id"
+Content-Type: text/plain
+
+123e4567-e89b-12d3-a456-426655440000
+--abcde12345
+Content-Disposition: form-data; name="profileImage"; filename="image1.png"
+Content-Type: image/png
+
+{…file content…}
+--abcde12345--
+```
+According to https://datatracker.ietf.org/doc/html/rfc7578, multipart request payload contains multi independent parts which could be divided into two kinds: one is non-file part which contains `required "name"/optional "Content-Type" / content`; the other one is file part which contains one more `optional "filename"`.
+
+# Current TCGC API for multipart
+Currently TCGC only has boolean flag [isMultipartFileInput](https://github.com/Azure/typespec-azure/blob/ab7a066d4ac0ae23a40f9ff8f4b6037559bda34c/packages/typespec-client-generator-core/src/interfaces.ts#L368) to distinguish file and non-file. After https://github.com/microsoft/typespec/issues/3046 complete, Typespec permit users to define more info explicitly(e.g. content-type) for each part so boolean flag is not enough.
+
+# Multipart in Typespec
+Typespec support two kinds of definition for multipart:
+
+1. Model format
+
+```
+op upload(
+  @header `content-type`: "multipart/form-data",
+  @multipartBody body: {
+    fullName: HttpPart<string>,
+    headShots: HttpPart<Image>[]
+  }
+): void;
+```
+
+2. Tuple format
+
+ ```
+op upload(
+  @header `content-type`: "multipart/form-data",
+  @multipartBody body: [
+    HttpPart<string, #{ name: "fullName" }>,
+    HttpPart<File, #{ name: "file" }>
+  ]
+): void;
+```
+
+# Proposal about new TCGC API for multipart
+Since all language emitters emit multipart body as model format, TCGC will uniform the multipart body type as model, and each model property equals to property of model format or part of tuple format in Typespec.
+
+```typescript
+exprot interface multipartOptionsType {
+  isFilePart: boolean; // whether this part is for file
+  multi: boolean; // whether this part is multi in request payload
+  headers: HeaderProperty[]; // relates to custom header
+  filename?: SdkModelPropertyTypeBase;
+  contentType?: SdkModelPropertyTypeBase;
+}
+
+export interface SdkBodyModelPropertyType extends SdkModelPropertyTypeBase {
+  kind: "property";
+  discriminator: boolean;
+  serializedName: string;
```

Comment:
`name` is not strictly necessary for non form-data content types such as `multipart/mixed`. Something like this could be in fact valid:

```typespec
op upload(
  @header `content-type`: "multipart/mixed",
  @multipartBody body: [
    HttpPart<string, #{}>,
    HttpPart<File, #{}>
  ]
): void;
```

Have we thought about how we might handle this?

Reply:
Nice hint. For multipart/mixed, the name is not needed so it is also not defined in Typespec. I add `isNameDefined` in multipartOptionsType so that language emitters could get the info there is no name defined.

Reply:
Is there only one Part be no name? Or multiple Parts is also valid?

I'd agree that if there is no name in Part, the `serializedName` should be undefined/empty, as this represents the wire data.

However, I am not sure whether TCGC would need to fill-in the `name` -- this would be similar to "anonymous model", when typespec does not have a name, TCGC would fill one.
Though, it seems had to figure out a proper name here.

If we won't see a `multipart/mixed` in azure anytime soon, we can have an empty `name`, and delay this discussion to future iterations.

Reply:
To be honest, I don't know much about `multipart/mixed` and I have not seen any real service adopt this kind of content-type. With `isNameDefined`, at least TCGC have provided the info "name is not defined in Typespec and language may have to decide by yourself how to expose this part". Maybe we could delay the discussion and focus on "multipart/form-data" at first.

Reply:
I am happy to delay this -- I mainly wanted to bring it up so that we're aware of it. My understanding is that part of the reason for the new TypeSpec API is to allow for different kinds of multipart payloads (not just form data) to be expressed so maybe this is something we will want to support in the future. But if this design for TCGC is focused just on multipart/form-data, I don't have a problem with that.

Using `multipart/mixed` in Azure seems to be rare but it's not unheard of -- as an example, the JS Tables SDK has some custom code handling the content type [here](https://github.com/Azure/azure-sdk-for-js/blob/af883f4bdb91b8d8aee15c45bd9eb85e0c9c351c/sdk/tables/data-tables/src/utils/transactionHelpers.ts).